addons/loyalty/models/product_product.py

```diff
@@ -14,3 +15,19 @@ def write(self, vals):
             if rewards:
                 raise ValidationError(_("This product may not be archived. It is being used for an active promotion program."))
         return super().write(vals)
+
+    @api.ondelete(at_uninstall=False)
+    def _unlink_except_loyalty_products(self):
+        product_data = [self.env.ref('loyalty.gift_card_product_50', False), self.env.ref('loyalty.ewallet_product_50', False)]
+        for product in self.filtered(lambda p: p in product_data):
+            raise UserError(_("You cannot delete %s as it is used in another module but you can archive it instead.", product.name))
+
+
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    @api.ondelete(at_uninstall=False)
+    def _unlink_except_loyalty_products(self):
+        product_data = [self.env.ref('loyalty.gift_card_product_50', False), self.env.ref('loyalty.ewallet_product_50', False)]
+        for product in self.filtered(lambda p: p.product_variant_id in product_data):
+            raise UserError(_("You cannot delete %s as it is used in another module but you can archive it instead.", product.name))
```

Comment:
Phrase "another module" doesn't make any sense for user. They just see Odoo UI and they don't know what is current module and what is another module